src/common/download.cpp

```diff
@@ -40,27 +41,96 @@
 
 namespace tools
 {
-  static bool download_thread(const std::string &path, const std::string &url)
+  class unlockable_lock_guard {
+  public:
+    unlockable_lock_guard(boost::mutex &m): mutex(m), locked(true) { mutex.lock(); }
+    ~unlockable_lock_guard() { if (locked) mutex.unlock(); }
+    void lock() { mutex.lock(); locked = true; }
+    void unlock() { mutex.unlock(); locked = false; }
+  private:
+    boost::mutex &mutex;
+    bool locked;
+  };
+
+  struct download_thread_control
+  {
+    const std::string path;
+    const std::string uri;
+    std::function<void(const std::string&, const std::string&, bool)> result_cb;
+    std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress_cb;
+    bool started;
+    bool stop;
+    bool stopped;
+    bool success;
+    boost::thread thread;
+    boost::mutex mutex;
+    boost::mutex started_lock;
+    boost::condition_variable started_cond;
+
+    download_thread_control(const std::string &path, const std::string &uri, std::function<void(const std::string&, const std::string&, bool)> result_cb, std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress_cb):
+        path(path), uri(uri), result_cb(result_cb), progress_cb(progress_cb), started(false), stop(false), stopped(false), success(false) {}
+    void lock() { mutex.lock(); }
+    void unlock() { mutex.unlock(); }
+  };
+
+  static void download_thread(download_async_handle control)
   {
+    static std::atomic<unsigned int> thread_id(0);
+
+    MLOG_SET_THREAD_NAME("DL" + std::to_string(thread_id++));
+
+    control->started_lock.lock();
+    control->started = true;
+    control->started_cond.notify_one();
+    control->started_lock.unlock();
+
     try
     {
-      MINFO("Downloading " << url << " to " << path);
+      unlockable_lock_guard lock(control->mutex);
```

Comment:
Ok, I'm sorry for being so irritating but I just remembered its 100% easier to do this in the destructor of `~download_thread_control() { if (thread.joinable()) thread.detach(); }`. Also, the way I suggested would be bad because the lock is not held on the `thread.join()` calls so pthread would get mad.

And perhaps comments should be added for the exposed functions doing the join calls - it is not safe to call those functions from two different threads.